File: mobile/src/main/kotlin/de/scout/fireplace/bus/RxBus.java
package de.scout.fireplace.bus;

import io.reactivex.Flowable;
import io.reactivex.processors.PublishProcessor;

public class RxBus {

  private static final PublishProcessor<Object> bus = PublishProcessor.create();
  //    private static final BehaviorSubject<Object> bus = BehaviorSubject.create();

  // If multiple threads are going to emit events to this
  // then it must be made thread-safe like this instead
  //    private final Subject<Object, Object> _bus = new SerializedSubject<>(PublishSubject.create());

  public void send(Object o) {
    bus.onNext(o);
  }

  public Flowable<Object> toObserverable() {
    return bus;
  }

  public boolean hasSubscribers() {
    return bus.hasSubscribers();
  }

  public static RxBus getInstance() {
    return new RxBus();
  }

  private RxBus() {
    // No instances.
  }
}

File: feature/src/main/kotlin/de/scout/fireplace/feature/bus/RxBus.kt
package de.scout.fireplace.feature.bus

import io.reactivex.processors.PublishProcessor

@Deprecated("Not thread safe")
internal class RxBus private constructor() {

  fun send(o: Any) = bus.onNext(o)

  fun toObservable() = bus!!

  companion object {

    private val bus = PublishProcessor.create<Any>()

    val instance: RxBus get() = RxBus()
  }
}
